sip/sip-007-stacking-consensus.md

```diff
@@ -534,21 +534,26 @@ Where `total_block_commit_amount` is equal to the sum of outputs [1, M+1].
 After the sunset phase _ends_ (i.e., blocks >= 500,000th burn block), block commits are _only_ burns, with
 a single burn output at index 1.
 
-## Stacking Operations on Bitcoin
+## STX Operations on Bitcoin
 
-As described above, PoX allows stackers to submit `stack-stx` operations on Bitcoin as well as on
-the Stacks blockchain. Any such operations must be evaluated by miners before beginning execution of
-any anchor block following those operations.
+As described above, PoX allows stackers to submit `stack-stx`
+operations on Bitcoin as well as on the Stacks blockchain. The Stacks
+chain also allows addresses to submit STX transfers on the Bitcoin
+chain. Such operations are only evaluated by the miner of an anchor block
+elected in the burn block that immediately follows the burn block that included the
+operations. For example, if a `TransferStxOp` occurs in burnchain block 100, then the
+Stacks block elected by burnchain block 101 will process that transfer.
```

Comment:
Is this as accurate as we can make it?  Aren't miners that mine on a different fork off block 100 also responsible to include the transfer in their block?  Seems like the code is doing this, but this should also be reflected here.

Reply:
There's no "fair" way to do this.  The fundamental problem is that there can be an unbound number of Bitcoin blocks, and thus an unbound number of on-Bitcoin operations to process, in-between any pair of Stacks blocks.  We can't process them all; otherwise, the network would be vulnerable to a DDoS attack if a miner simply mines a Stacks block, waits for a long time (e.g. 10,000 Bitcoin blocks), and then mines a child off of that Stacks block.  Without constraining which Bitcoin blocks' on-chain operations will be considered, each node on the network network would be obliged to consider _all_ on-Bitcoin operations in those 10,000 Bitcoin blocks (even the act of iterating through them would be time-consuming).

The act of only considering on-chain Bitcoin operations from the last Bitcoin block is one of many possible ways to deal with this problem, and was chosen because it's the easiest to communicate to users.  Since miners will build off of the longest Stacks chain most of the time, and will try to mine blocks in each Bitcoin block most of the time, we can explain to users that if they send an on-Bitcoin STX-transfer, it'll get picked up by the Stacks miner who mines in the next Bitcoin block (with the caveats that there may be no block mined, or that the block mined might get orphaned; in both cases, the remedy is to try again).  Most of the time, this would work -- the user experience would be equivalent to sending a stx-transfer transaction and having it get mined in the next Stacks block.

Reply:
given the cost of a btc transaction, I don't think we have to worry much about having too many transactions used as a DDoS vector.  I'm actually a lot more worried of people having to spend several dollars to move only a few hundred STX (when a lot of them have been told to wait for Stacks 2.0 where the fees will be cheaper), although I'm not clear how many people will be affected as I'm not sure whether that's anyone with STX prior to 2.0, or only some using a certain wallet.

My understanding of the code (https://github.com/blockstack/stacks-blockchain/pull/2129/files#diff-36b251011e9d6834be636547f23dd28aa9b24b555331f637dd40cfd2a1552467R3790) is that it will try to apply all stx transfers from the last pox anchor block if it has not been already included in the current fork, and that behavior should be reflected in the SIP.  Same logic used to consider stacking via BTC.  But maybe I misunderstood the code, some times it is hard to read just from diffs....

Reply:
> given the cost of a btc transaction, I don't think we have to worry much about having too many transactions used as a DDoS vector.

We _absolutely_ have to worry about this.  Per my above comment, the number of on-BTC operations between any parent/child Stacks block is unbound -- it can be arbitrarily high, since the number of BTC blocks in-between any parent/child Stacks block can be arbitrarily high (since I can mine a Stacks block off of any existing Stacks block).  It may not be a problem right now, but that does not mean it won't ever be a problem.  In fact, the "size" of the problem increases with the length of the Bitcoin chain -- even if there are no on-BTC operations in-between a parent/child Stacks block pair, the processing logic still needs to iterate through all of the Bitcoin blocks in-between these two Stacks blocks' sortitions (which grows unbound with the length of the Bitcoin chain).

> My understanding of the code is that it will try to apply all stx transfers from the last pox anchor block if it has not been already included in the current fork, and that behavior should be reflected in the SIP

It applies all STX transfers from the last Bitcoin block, not the last PoX anchor block.

Reply:
Yep -- the place where the stacks transfers to be processed are selected is roughly here: https://github.com/blockstack/stacks-blockchain/pull/2129/files#diff-36b251011e9d6834be636547f23dd28aa9b24b555331f637dd40cfd2a1552467R4014 -- the block processor fetches all the valid transfer operations from the parent burn block, and those are the ones later passed into the function you linked. So that behavior is pretty much as described in the SIP. The only way a Stacks transfer broadcasted on Bitcoin may be processed in two different forks is if there's a Bitcoin fork or a PoX fork.

> given the cost of a btc transaction, I don't think we have to worry much about having too many transactions used as a DDoS vector. I'm actually a lot more worried of people having to spend several dollars to move only a few hundred STX (when a lot of them have been told to wait for Stacks 2.0 where the fees will be cheaper), although I'm not clear how many people will be affected as I'm not sure whether that's anyone with STX prior to 2.0, or only some using a certain wallet.

The reason that it becomes a DDoS vector is that it's somewhat "unbound" how much processing would need to be done if the Stacks node processed all the on-Bitcoin operations between it's Stacks block parent and the current block. Imagine this scenario:

```
Bitcoin blocks:    1 -> 2 -> 3 -> 4 -> ... -> 10,000
```

And those Bitcoin blocks "elect" the following Stacks blocks:

```
Stacks blocks:     a -> b -> c -> d -> ...
                    \---------------------> b'
```
where `b'` is elected in Bitcoin block 10,000. Now, if there's 5 STX transfers in every Bitcoin block from 2 -> 10,000, then each Stacks node needs to process 50,000 STX transfers right before processing `b'`. On top of that, this DDoS vector never goes away -- a subsequent miner could mine `c'` as a child of `b` and nodes would need to perform that processing again.

Reply:
thank you for the explanation @kantai 

I just can't get over the fact the finality of the transaction is not guaranteed.  There has to be a way to make this work like the mempool works for regular Stacks transactions and guarantee all Stacks forks get the transaction.

Reply:
Yeah -- it could work that way if we implemented some of the items discussed in #2094. In particular, if the `stacks-node` could validate Bitcoin signatures itself without relying on bitcoind (this would be something like a parallel implementation of bitcoin signature verification), then these transactions could be considered "legal" stacks transactions in their own right.